actions/setup-nix/action.yml

```diff
@@ -1,107 +1,159 @@
           exit 1
         fi
 
-        # symlink credential folders for nix build access
-        # nix build is run via daemon on root, credentials linked for daemon access
+        echo "🔗 Symlinking AWS credentials for Nix daemon..."
         sudo ln -s ~/.aws $ROOT_PATH
 
-    - uses: cachix/install-nix-action@ba0dd844c9180cbf77aa72a116d6fbc515d0e87b # v27
+    - name: 💾 Store private signing key (if provided)
+      if: ${{ inputs.private-signing-key != '' }}
+      shell: bash
+      run: |
+        echo "🔑 Storing private signing key for cache signing..."
+        echo "${{ inputs.private-signing-key }}" > ${{ github.workspace }}/nix-signing.private
+        chmod 600 ${{ github.workspace }}/nix-signing.private
+
+    - name: Preprocess Public Keys
+      shell: bash
+      run: |
+        # Preprocess the public keys and store them in a variable
+        PUBLIC_KEYS=$(echo "${{ inputs.cache-pubkeys }}" | tr ',' ' ')
+        echo "Processed Public Keys: $PUBLIC_KEYS"
+
+        # Store the processed public keys in an environment variable for use in later steps
+        echo "PUBLIC_KEYS=$PUBLIC_KEYS" >> $GITHUB_ENV
+
+    - uses: DeterminateSystems/nix-installer-action@da36cb69b1c3247ad7a1f931ebfd954a1105ef14 # v14
       with:
-        install_url: https://releases.nixos.org/nix/nix-2.18.1/install
-        nix_path: nixpkgs=channel:nixos-unstable
-        # Enables sandbox for all drv except those with `__noChroot = true;`
-        # Add caches for public nixos cache + public community cache + specified private cache
-        extra_nix_config: |
+        determinate: true
+        extra-conf: |
           sandbox = relaxed
-          substituters = https://cache.nixos.org https://nix-community.cachix.org ${{ inputs.cache-url }}
-          trusted-public-keys = cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY= nix-community.cachix.org-1:mB9FSh9qf2dCimDSUo8Zy7bkq5CX+/rkCWyvRCYg3Fs= ${{ inputs.cache-pubkey }}
+          substituters = https://cache.nixos.org ${{ inputs.cache-url }}?scheme=https&region=${{ inputs.aws-region }}
+          trusted-public-keys = cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY= ${{ env.PUBLIC_KEYS }}
+          secret-key-files = ${{ github.workspace }}/nix-signing.private
+
+    - name: ⚙️ Configure Nix for cache usage by modifying the script with `sed`
+      shell: bash
+      run: |
+        if [ "${{ inputs.cache-mode }}" == "push" ]; then
+          echo "🚀 Enabling push mode: configuring post-build hook for cache upload"
+
+          # Create a modified version of the script with the actual inputs for CACHE_URL and AWS_REGION
+          MODIFIED_SCRIPT_PATH="/tmp/upload-to-cache-modified.sh"
+          cp ${GITHUB_ACTION_PATH}/scripts/upload-to-cache.sh $MODIFIED_SCRIPT_PATH
+
+          # Use sed to replace the placeholders in the script
+          sed -i "s|export CACHE_URL=\$1|export CACHE_URL=${{ inputs.cache-url }}|" $MODIFIED_SCRIPT_PATH
```

Comment:
Hmm, do we really need to modify the script.

Could we use bash args in the `post-build-hook` or set an env variables before running the script?

Reply:
unfortunately the post build hook does not support passing params. you can only pass the script path. Env variables could be an option but using the composite actions we are not able to pass env variables here. We would need to rely on the user knowing the env variable is setup before calling the action.

Reply:
Couldn't the env var be set up in the action itself?

The input params to the action are both the AWS_REGION and CACHE_URL. Not sure if we can map this input to env vars for the action/job.

Reply:
I just tried this https://github.com/smartcontractkit/.github/pull/502/commits/e1678082bd30b8ef19031eb6abd3b8c8a739ee3d , but unfortunately it seems that after setting the env var value in the composite action and the composite action returning to the main workflow, the value gets resetted and subsequent uploads on the workflow fail: https://github.com/smartcontractkit/gauntlet-plus-plus/actions/runs/11145685297/job/30975857353?pr=515#step:4:2997

I will keep the original approach althoguh it's a bit more ugly.